### ISSUE TITLE
Store inconsistencies

### ISSUE THREAD

**Issue by torbensky** [battlesnakeio/roadmap#102]
https://github.com/battlesnakeio/engine/blob/master/controller/store.go#L34

`Lock` says the token is required to write frames for a game. However, the `PushGameFrame` method makes no use of a token. Furthermore, other game mutating operations like `CreateGame` and `SetGameStatus` also do not make use of a lock token.

1. Lock method dictates how game frame method works
2. Game frame method does not use the lock token in the way that the Lock method's documentation says it should

Recommended solution: Just change the documentation for `Lock` and leave the use of the `Lock` for a higher layer like the controller layer.

TLDR;
Update `Lock` documentation to not say writing frames requires a token. Leave other methods as-is.

**Comment by dlsteuer** [comment]
I think this might actually be a deficiency of the other methods and they probably should require a lock token to write game data. So I think we should actually fix the other methods and you can only mutate game data if you hold the lock.

**Comment by torbensky** [comment]
It's a poor architecture to push the locking down into the storage imo. That pushes a lot of complexity into the data layer, of which we already have 3 implementations. This would make the project much harder to maintain and change in the future. The locking should be done at the controller level. I think it will require a lot more work in the long run. Maybe even in the short term. I think the reason this hasn't been noticed before is because previous store implementations have been single process and not distributed and have access to things like local semaphores.

**Comment by ColDog** [comment]
Couple things, the original architecture was to not allow mutating game state if a lock is not held, I think @dlsteuer is right need to update those methods that are mutating state without holding locks.

@torbensky it's not necessarily a "poor architecture" it was designed to allow the storage interfaces to make optimizations as necessary for locking (a redis pipeline is a perfect example). Maybe provide some examples of how this could be cleaned up to help me out?

**Comment by torbensky** [comment]
Coordination shouldn't be pushed into the storage layer. There are many storage backends which don't have good support for distributed locking. But not only that, even if we restrict ourselves to the storage technologies that can do locking, we'd have to re-implement and test distributed locking in each persistence layer we use (which already is at 3).

Instead, we should have a separate coordination system that is independent of the storage layer. It's a pretty common practice in web applications to have a coordination separated out like this.

**Comment by brandonb927** [comment]
Just pinging for an update here, any resolution?

**Comment by dlsteuer** [comment]
Going to close this, these changes aren't going to be made right now in the engine.
